### PR TITLE
Extend cli commands with plugins

### PR DESCRIPTION
**Pull request by jsenin**
Added a search for plugins modules that contains a `def cli_commands()` that returns a clic.Command object. This object is merged with others commands to populate the cli.

An example file that should be placed at `~/.piku/plugins/postgres/__init__.py`

```
import click

@click.group()
def postgres():
    """Postgres command plugin"""
    pass

@postgres.command("postgres:create")
@click.argument('name')
@click.argument('user')
@click.argument('password')
def postgres_create():
    """Postgres create a database"""
    pass

@postgres.command("postgres:drop")
@click.argument('name')
def postgres_drop():
    """Postgres drops a database"""
    pass

@postgres.command("postgres:import")
@click.argument('name')
def postgres_drop():
    """Postgres import a database"""
    pass

@postgres.command("postgres:dump")
@click.argument('name')
def postgres_drop():
    """Postgres dumps a database SQL"""
    pass

def cli_commands():
    return postgres
```

### CODE REVIEW

**Comment by rcarmo** [comment]
I _really_ like this. @chr15m, what say ye? Before we merge (which I don't have a problem with, only the usual self-consciousness about not having set up testing yet...), can we use this for a brief discussion about what else we can do via plugins? Maybe even move some runtime support to plugins?

**Comment by jsenin** [comment]
I also missed tests, and I didn't want to start testing because choosing the test suite seems like a tough battle for a newcomer. Still I propose to use mamba (http://nestorsalceda.com/mamba/). I am really happy to use this suite, it is very comfortable.
About plugins, I thought about creating a 'core' or 'piku' plugin and moving all that piece of code to the plugin. But the question arises as to why the code is all in the piku.py file and not in modules. I guess it's for convenience, to copy a single file across servers instead directories, etc.., but right now it's big enough to think about it, or at least group code into classes.

I'm starting to implement some postgres commands because I need to create databases and restore sql dumps, in a easy way. I hope you like this improvement.

**Comment by rcarmo** [comment]
Thanks. I like the idea of plugins, since Piku is minimalist by design for the reasons you mentioned but cannot be all things to all people. The challenge is to figure out if overly complex language support (and which ones) should become plugins or if we should use them for setting up databases, certificates, message queues, etc.

**Comment by chr15m** [comment]
@rcarmo @jsenin I also really like this idea and the patch looks good with some minor tweaks.

To play devil's advocate for a moment, with regards to the PostgreSQL example, what would be the advantage of this over the user simply doing `piku run createdb` and using `release` workers to provision a database? For example in my own deployments I am using the following bash script as a release worker which idempotently creates the database:

```
#!/bin/sh
DATABASE=${DATABASE:-testdb}
psql "${DATABASE}" -c ";" 2>/dev/null && echo "Database ${DATABASE} already exists." || ( createdb "${DATABASE}" && echo "Database ${DATABASE} created." )
```

Already with the `release` and scripted deployments you can do deployments of e.g. Go binaries without plugins (as per the Gitea deploy example). I like the plugin idea for the same reason as it gives the user more flexibility and allows us to keep the core minimal.

File: piku.py [hunk 1]
```diff
@@ -1354,6 +1358,23 @@ def cmd_git_receive_pack(app):
     # Handle the actual receive. We'll be called with 'git-hook' after it happens
     call('git-shell -c "{}" '.format(argv[1] + " '{}'".format(app)), cwd=GIT_ROOT, shell=True)
 
+import importlib
```
Comment (chr15m) [inline piku.py]: Existing style is to have imports at the top.

File: piku.py [hunk 2]
```diff
@@ -1354,6 +1358,23 @@ def cmd_git_receive_pack(app):
...
+def _get_plugin_commands(path):
+    sys.path.append(os.path.abspath(path))
```
Comment (chr15m) [inline piku.py]: Please follow the style of the rest of the code which is to import the function you want and use it directly (e.g. `from os.path import abspath` and then `abspath()`. Same for `listdir` and `import_module` below.

Comment (jsenin) [reply piku.py]: yes of course, but @chr15m how do you transform `sys.path.append` ?

Comment (chr15m) [reply piku.py]: Good question. Maybe leave that one as `sys.path.append` or maybe @rcarmo has a better idea (he established the practice of importing e.g. `abspath`).

**Comment by jsenin** [comment]
> @rcarmo @jsenin I also really like this idea and the patch looks good with some minor tweaks.
> 
> To play devil's advocate for a moment, with regards to the PostgreSQL example, what would be the advantage of this over the user simply doing `piku run createdb` and using `release` workers to provision a database? For example in my own deployments I am using the following bash script as a release worker which idempotently creates the database:
> 
> ```
> #!/bin/sh
> DATABASE=${DATABASE:-testdb}
> psql "${DATABASE}" -c ";" 2>/dev/null && echo "Database ${DATABASE} already exists." || ( createdb "${DATABASE}" && echo "Database ${DATABASE} created." )
> ```
> 
> Already with the `release` and scripted deployments you can do deployments of e.g. Go binaries without plugins (as per the Gitea deploy example). I like the plugin idea for the same reason as it gives the user more flexibility and allows us to keep the core minimal.

My argumentation is the next one :
Using plugins like postgres has some advantages:

- first is a documented cli. Even if i dont read documentation i can see that i can operate with a postgres database, needed arguments, etc...
- easy management. It's more easy to use a wrapper that have knownledge about internal or specialized commands, limitations and settings for each technology than I remember or memorize it. 
- easy adoption. Users that already use dokku or heroku cli's could find the similar commands, and start to use piku easily 

For the postresql, advandages that I see are:
- easy database manipulation, psql commands are not needed
- do not need to deploy. Manage database witouth deploying
- do not need to code. No code required to do stuff with the database

About 'release' worker I like it, but I think that solve some concrete needed for the application but not for the operator, developer, user, etc..

File: piku.py (hunk 2 above)
Comment (rcarmo) [reply piku.py]: I would leave it as sys.path.append for now (I like to import individual functions for conciseness and less namespace junk, but stuff with common names is... tricky). Maybe import it as `path_append` for readability, but I’m good either way...

**Comment by rcarmo** [comment]
Okay, so here’s an idea - we deal with #128 and #130 before this one and cut it as  the 1.0 release (so that we can do self-updates, etc.) and then get back to this, which I think has a lot of potential.

Maybe do a 0.9 to test self updates first.

Opinions?

**Comment by jsenin** [comment]
Sounds great !

El 18 de noviembre de 2019 22:41:54 CET, Rui Carmo <notifications@github.com> escribió:
>Okay, so here’s an idea - we deal with #128 and #130 before this one
>and cut it as  the 1.0 release (so that we can do self-updates, etc.)
>and then get back to this, which I think has a lot of potential.>
>>
>Maybe do a 0.9 to test self updates first.>
>>
>Opinions?>
>>
>-- >
>You are receiving this because you were mentioned.>
>Reply to this email directly or view it on GitHub:>
>https://github.com/piku/piku/pull/129#issuecomment-555222445

-- 
Enviado desde mi dispositivo Android con K-9 Mail. Por favor, disculpa mi brevedad.

**Comment by dwightmulcahy** [comment]
Yeah, I'm liking the idea of this one too!

**Comment by chr15m** [comment]
:ok_hand: :+1:

**Comment by rcarmo** [comment]
Since we’re all here discussing this one, here’s my gut feeling about what should be kept in core or moved to plugins (just a tentative list, subject to change):

## Core:

* Python (possibly including pyenv) and Node support (because that’s what I see being used in schools, and are a good fit for a vanilla install on a Pi)
* Basic job scheduling, scaling, etc.
* TCP Port management for stuff that isn’t run through `nginx`
* Basic logging (pretty much as-is, my biggest concern with that was making sure there was log rotation and that beginners wouldn’t stuff a Pi full of log files)

I think we can keep this in a single file

## Plugins:

* Java/Clojure, Go, Lua, Elixir, Docker/Podman, etc. - especially things that require convoluted build steps that may add unnecessary complexity to core
* _maybe_ SSL (Let’s Encrypt, etc., although self-signed certificates are so trivial that they can probably stay in core)
* Database management (pg, Redis, etc.) 
* Logging and metrics (I’d like to have hooks to send out data to external systems)

The basic notion here is that even though we can use Piku for “big stuff” like running public services, the core should be something that is immediately useful for hobbyists and K12 to get something running quickly.

What do you guys think?

**Comment by chr15m** [comment]
@jsenin I guess you didn't run this code in its current state as `os.path.is_dir` is not a valid import.

**Comment by chr15m** [comment]
@rcarmo to me personally this list looks good and makes sense. Will be nice to get the file size back down.

I would be sad to lose acme.sh/SSL support from core. For me the ability to bootstrap a domain name onto a running app with no configuration at all is a killer feature. But if I can tell people how to get it working without too many commands then all good.

**Comment by rcarmo** [comment]
SSL is just an example because it’s a sizable dependency (outside Piku). I’m fine with it staying in core depending on LOC :)

**Comment by jsenin** [comment]
> @jsenin I guess you didn't run this code in its current state as `os.path.is_dir` is not a valid import.

fixed

**Comment by jsenin** [comment]
> SSL is just an example because it’s a sizable dependency (outside Piku). I’m fine with it staying in core depending on LOC :)
> […](#)
> On 19 Nov 2019, at 09:06, Chris McCormick ***@***.***> wrote: ﻿ @rcarmo to me personally this list looks good and makes sense. Will be nice to get the file size back down. I would be sad to lose acme.sh/SSL support from core. For me the ability to bootstrap a domain name onto a running app with no configuration at all is a killer feature. But if I can tell people how to get it working without too many commands then all good. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or unsubscribe.

I'm agree leaving SSL at core, and your proposal @rcarmo .

> The basic notion here is that even though we can use Piku for “big stuff” like running public services, the core should be something that is immediately useful for hobbyists and K12 to get something running quickly.

what's k12 ?

**Comment by dwightmulcahy** [comment]
>the core should be something that is immediately useful for hobbyists and K12 to get something running quickly.

@rcarmo I agree with you about that.

@jsenin 
K12 means "Kindergarten thru 12 grade"

**Comment by chr15m** [comment]
@jsenin looks like this PR needs master merged in.

**Comment by chr15m** [comment]
@jsenin I've sent you a PR to pull this into line against master and also fix a couple of bugs I found when running it on my machine.

**Comment by chr15m** [comment]
@jsenin I pushed another PR against your branch to clean up the imports a bit so they have less impact.

https://github.com/jsenin/piku/pull/4

**Comment by chr15m** [comment]
@jsenin my PR is broken, please do not merge yet. :man_facepalming: Glad @rcarmo added the linting test!

**Comment by chr15m** [comment]
@jsenin ok I finally got a PR working that cleanly merges against master and passes the linting and run test.

If you merge https://github.com/jsenin/piku/pull/4 into your piku-cli-plugins then this should be good to go.

**Comment by rcarmo** [comment]
I'm going to merge this since it's both long overdue and a clean way for people to expand `piku`.

**Pull request merged**